### PR TITLE
SPARK-21220 Use outputPartitioning as bucketing if possible

### PR DESCRIPTION
**Pull request by ash211**
## What changes were proposed in this pull request?

If possible, set the bucketing columns to be the outputPartitioning if it's a HashPartitioning

## How was this patch tested?

TODO

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 1]
```diff
@@ -122,7 +122,14 @@ object FileFormatWriter extends Logging {
       // guarantee the data distribution is same between shuffle and bucketed data source, which
       // enables us to only shuffle one side when join a bucketed table and a normal one.
       HashPartitioning(bucketColumns, spec.numBuckets).partitionIdExpression
-    }
+    }.orElse(
+      // If there are no columns in the bucket spec, check if the data is already HashPartitioned
+      // and use that HashPartitioning's partitionIdExpression
+      plan.outputPartitioning match {
```
Comment (robert3005) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: Sorry, had to http://blog.originate.com/blog/2014/06/15/idiomatic-scala-your-options-do-not-match/.

Spark prefers map.getOrElse.

Comment (robert3005) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala]: I see, this is not an option. I think you will need to rebind expression here looking at test failures and also add a filter

**Comment by pwoody** [comment]
Probably want this for sort order as well?

**Comment by robert3005** [comment]
@ash211 ping?

**Comment by robert3005** [comment]
Reopen if you're going to update it

**Pull request closed without merging**